netket/experimental/operator/_fermion_operator_2nd.py

```diff
@@ -476,25 +516,31 @@ def _op__imatmul__(self, other):
 
         terms = []
         weights = []
-        for t, w in self._operators.items():
-            for to, wo in other._operators.items():
+        for t, w in zip(self.terms, self.weights):
+            for to, wo in zip(other.terms, other.weights):
                 terms.append(tuple(t) + tuple(to))
                 weights.append(w * wo)
-        if not _isclose(other._constant, 0.0):
-            for t, w in self._operators.items():
+        if not np.isclose(other._constant, 0.0):
+            for t, w in zip(self.terms, self.weights):
                 terms.append(tuple(t))
                 weights.append(w * other._constant)
-        if not _isclose(self._constant, 0.0):
-            for t, w in other._operators.items():
+        if not np.isclose(self._constant, 0.0):
+            for t, w in zip(other.terms, other.weights):
                 terms.append(tuple(t))
                 weights.append(w * self._constant)
         constant = self._constant * other._constant
 
-        self._operators = _remove_dict_zeros(dict(zip(terms, weights)))
+        self._terms = terms
+        self._weights = weights
         self._constant = constant
         self._reset_caches()
         return self
 
+    def reduce(self):
+        """Prunes the operator by removing all terms with zero weights (inplace)"""
+        self._terms, self._weights = _remove_zero_weights(self._terms, self._weights)
```

Comment:
I think reduce should also convert to normal order (or another ordering) in order to remove duplicates (you could have two equivalent terms which are the same but since the operator is unordered it does not notice)